docs/src/cli/deploy-a-program.md

```diff
@@ -297,6 +308,74 @@ the program:
+
+### Upgrading program using offline signer as authority
+
+Storing private key(s) on machine without internet access (offline signer) is much safer compared to
+storing them on machine with internet access (online signer), this section describes how program developer
+can use offline machine (offline signer) to sign-off on his program `upgrade` to get a **higher degree of
+security** (compared to typical upgrade flow described in [this section](deploy-a-program.md#redeploy-a-program)
+which assumes the usage of machine connected to internet - aka online signer). Note, currently only `upgrade`
+operation can be performed in "offline" mode, initial program `deploy` **must** be performed using online machine
+and only subsequent program upgrades can leverage offline signing. For a first time deploy it is **recommended** 
+to [change program's upgrade authority (using `--skip-new-upgrade-authority-signer-check` option)](deploy-a-program.md#set-a-programs-upgrade-authority)
+to offline signer, so that future program upgrades will only be possible with a signature of private key residing on 
+offline machine.
+
+Assuming your program has been deployed (using online machine) and its upgrade authority has been changed to
+offline signer (a signer whose private key is generated and stays on machine **disconnected** from internet), 
+a typical setup would consist of 3 different signers (pairs of keys):
+- online signer (used as fee payer for deploying program buffer, deploying/upgrading program itself)
+- offline signer (serves as authority over program upgrades, protects program upgrades from certain 
+  types of attacks), make sure to have a **secure** backup of this signer
+- program signer (typically, generated when program has been compiled into `.so` file)
+
+The general process is as follows:
+1) (use online machine) create buffer
+2) (use online machine) upgrade buffer authority to offline signer
+3) (optional, use a separate online machine) verify the actual buffer on-chain contents
+4) (use offline machine) build and copy a signature for your intent to upgrade program
+5) (use online machine) use this signature to build and broadcast upgrade transactions on-chain
+
+```bash
+# (1) (use online machine) create buffer
+solana program write-buffer <PROGRAM_FILEPATH>
+
+# (2) (use online machine) upgrade buffer authority to offline signer
+solana program set-buffer-authority <ONLINE_SIGNER_PUB_KEY> --new-buffer-authority <OFFLINE_SIGNER_PUB_KEY>
+```
+
+- note, above we have to issue a separate `solana program set-buffer-authority` command (using online machine) to 
+  change authority to be that of offline signer because we don't have access to offline signer private key to sign 
+  up as buffer authority when running `solana program write-buffer` command using online machine (offline signer 
+  private key is stored on offline machine), and `solana program set-buffer-authority` doesn't impose that requirement.
```

Comment:
Yeah, this was an attempt to clarify limitations of `solana program write-buffer` - which I found confusing (as I mentioned in https://github.com/solana-labs/solana/pull/33860#discussion_r1373460162 above),

essentially this answers potential User question of "why do I need a separate command for setting the desired (offline) authority on the buffer I've created ? I'd rather set it to desired authority when creating buffer".

This might not be the best place to describe this caveat though, do you think we should:
- not mention it at all ?
- resolve it by extending `solana program write-buffer` command ?
- move this description somewhere else (perhaps under section that describes `solana program write-buffer` command) ?

Reply:
> Let's keep it simple and add a line in the write-buffer section that explains that it's not currently possible to set the buffer authority during creation.

I've moved these changes into separate PR https://github.com/solana-labs/solana/pull/34366, removed from this one in https://github.com/solana-labs/solana/pull/33860/commits/63e6e02b4199b1eeabfae15966138e039b3add06